rcl_action/src/rcl_action/action_server.c

```diff
@@ -0,0 +1,759 @@
+      action_server->impl->goal_handles[i], &status_message->msg.status_list.data[i].status);
+    if (RCL_RET_OK != ret) {
+      return RCL_RET_ERROR;
+    }
+  }
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rcl_action_publish_status(
+  const rcl_action_server_t * action_server,
+  const void * status_message)
+{
+  if (!rcl_action_server_is_valid(action_server)) {
+    return RCL_RET_ACTION_SERVER_INVALID;  // error already set
+  }
+  RCL_CHECK_ARGUMENT_FOR_NULL(status_message, RCL_RET_INVALID_ARGUMENT);
+
+  rcl_ret_t ret = rcl_publish(&action_server->impl->status_publisher, status_message);
+  if (RCL_RET_OK != ret) {
+    return RCL_RET_ERROR;  // error already set
+  }
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rcl_action_take_result_request(
+  const rcl_action_server_t * action_server,
+  void * ros_result_request)
+{
+  TAKE_SERVICE_REQUEST(result);
+}
+
+rcl_ret_t
+rcl_action_send_result_response(
+  const rcl_action_server_t * action_server,
+  void * ros_result_response)
+{
+  SEND_SERVICE_RESPONSE(result);
+}
+
+rcl_ret_t
+rcl_action_clear_expired_goals(
+  const rcl_action_server_t * action_server,
+  size_t * num_expired)
+{
+  if (!rcl_action_server_is_valid(action_server)) {
+    return RCL_RET_ACTION_SERVER_INVALID;
+  }
+  RCL_CHECK_ARGUMENT_FOR_NULL(num_expired, RCL_RET_INVALID_ARGUMENT);
+
+  // Get current time (nanosec)
+  int64_t current_time;
+  rcl_ret_t ret = rcl_clock_get_now(&action_server->impl->clock, &current_time);
+  if (RCL_RET_OK != ret) {
+    return RCL_RET_ERROR;
+  }
+
+  // Used for shrinking goal handle array
+  rcl_allocator_t allocator = action_server->impl->options.allocator;
+
+  *num_expired = 0u;
+  rcl_ret_t ret_final = RCL_RET_OK;
+  const int64_t timeout = (int64_t)action_server->impl->options.result_timeout.nanoseconds;
+  rcl_action_goal_handle_t * goal_handle;
+  rcl_action_goal_info_t goal_info;
+  int64_t goal_time;
+  size_t num_goal_handles = action_server->impl->num_goal_handles;
+  for (size_t i = 0; i < num_goal_handles; ++i) {
+    goal_handle = action_server->impl->goal_handles[i];
+    // Expiration only applys to terminated goals
+    if (rcl_action_goal_handle_is_active(goal_handle)) {
+      continue;
+    }
+    ret = rcl_action_goal_handle_get_info(goal_handle, &goal_info);
+    assert(RCL_RET_OK == ret);
+    goal_time = _goal_info_stamp_to_nanosec(&goal_info);
+    assert(current_time > goal_time);
+    if ((current_time - goal_time) > timeout) {
+      // Stop tracking goal handle
+      ret = rcl_action_goal_handle_fini(goal_handle);
+      if (RCL_RET_OK != ret) {
+        ret_final = RCL_RET_ERROR;
+      }
+      // Fill in any gaps left in the array with pointers from the end
+      action_server->impl->goal_handles[i] = action_server->impl->goal_handles[num_goal_handles];
+      action_server->impl->goal_handles[num_goal_handles--] = NULL;
+
+      ++(*num_expired);
+    }
+  }
+
+  // Shrink goal handle array if some goals expired
+  if (*num_expired > 0u) {
```

Comment:
This should deallocate if the new size is zero because `allocator.reallocate()` [is supposed to behave like realloc()](https://github.com/ros2/rcutils/blob/3c982d048914e1c30002d8ea69d86ac1cc6976fd/include/rcutils/allocator.h#L58-L62) which has [undefined behavior when the new size is zero](https://en.cppreference.com/w/c/memory/realloc).

Reply:
Addressed in c6b436e